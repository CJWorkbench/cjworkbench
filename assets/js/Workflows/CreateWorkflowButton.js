import React, { useState } from 'react'
import { csrfToken } from '../utils'

function CreateWorkflowButton ({ children }) {
  const [isSubmitted, setSubmitted] = useState(false)

  return (
    <form className='create-workflow' method='post' action='/workflows' onSubmit={() => setSubmitted(true)}>
      <input type='hidden' name='csrfmiddlewaretoken' value={csrfToken} />
      <button type='submit' disabled={isSubmitted}>
        {children}
      </button>
    </form>
  )
}
export default CreateWorkflowButton
