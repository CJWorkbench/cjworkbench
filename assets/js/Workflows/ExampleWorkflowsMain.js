import { Trans } from '@lingui/macro'
import WorkflowList, { WorkflowListPropType } from './WorkflowList'

export default function ExampleWorkflowsMain (props) {
  const { workflows } = props
  return (
    <main className='workflows'>
      <header>
        <h1><Trans id='js.Workflows.WorkflowLists.nav.recipes'>Example workflows</Trans></h1>
      </header>
      {workflows.length === 0 ? (
        <div className='placeholder'>
          <Trans id='js.Workflows.WorkflowLists.publishNewRecipes'>Publish workflows as examples using Django admin</Trans>
        </div>
      ) : (
        <WorkflowList className='example' workflows={workflows} />
      )}
    </main>
  )
}
ExampleWorkflowsMain.propTypes = {
  workflows: WorkflowListPropType.isRequired
}
