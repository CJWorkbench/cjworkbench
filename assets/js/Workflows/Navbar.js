import React from 'react'
import PropTypes from 'prop-types'
import WfHamburgerMenu from '../WfHamburgerMenu'
import { Trans } from '@lingui/macro'

function getActiveSection () {
  return window.location.pathname.startsWith('/workflows') ? 'workflows' : 'lessons'
}

function propsForLink (activeSection, name) {
  if (name === activeSection) {
    return { href: null, className: 'active' }
  } else {
    return { href: `/${name}`, className: 'inactive' }
  }
}

export default class Navbar extends React.Component {
  static propTypes = {
    user: PropTypes.shape({ id: PropTypes.number.isRequired }) // null/undefined if logged out
  }

  render () {
    const activeSection = getActiveSection()
    const { user } = this.props

    return (
      <div>
        <nav className='navbar'>
          <div className='logo'>
            <img src={`${window.STATIC_URL}images/logo.svg`} className='logo' />
            <img src={`${window.STATIC_URL}images/logo-text-dark.svg`} className='logo-text' />
          </div>
          <div className='links'>
            <a {...propsForLink(activeSection, 'workflows')}>
              <Trans id='js.Workflows.NavBar.links.workflows' description='This is used in navigation bar. It should be all-caps for styling reasons.'>WORKFLOWS</Trans>
            </a>
            <a {...propsForLink(activeSection, 'lessons')}>
              <Trans id='js.Workflows.NavBar.links.training' description='This is used in navigation bar. It should be all-caps for styling reasons.'>TRAINING</Trans>
            </a>
          </div>
          <WfHamburgerMenu user={user} />
        </nav>
      </div>
    )
  }
}
