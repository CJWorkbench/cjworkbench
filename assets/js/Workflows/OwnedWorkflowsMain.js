/* globals confirm */
import React from 'react'
import PropTypes from 'prop-types'
import { Trans, t } from '@lingui/macro'
import WorkflowList, { WorkflowListPropType } from './WorkflowList'
import CreateWorkflowButton from './CreateWorkflowButton'

function beginEdit (edits, id, changes) {
  const edit = edits[id]
    ? { changes: { ...edits[id].changes, ...changes }, nPendingChanges: edits[id].nPendingChanges + 1 }
    : { changes, nPendingChanges: 1 }
  return { ...edits, [id]: edit }
}

function endEdit (edits, id) {
  const edit = { ...edits[id], nPendingChanges: edits[id].nPendingChanges - 1 }
  return { ...edits, [id]: edit }
}

function reduceWorkflowEdit (state, action) {
  const { adds, edits } = state
  switch (action.type) {
    case 'begin-edit': {
      const { id, changes } = action.payload
      return { adds, edits: beginEdit(edits, id, changes) }
    }
    case 'end-edit': {
      const { id } = action.payload
      return { adds, edits: endEdit(edits, id) }
    }
    case 'begin-duplicate': {
      const { id } = action.payload
      return { adds, edits: beginEdit(edits, id, {}) }
    }
    case 'end-duplicate': {
      const { id, workflow } = action.payload
      return { adds: [workflow, ...adds], edits: endEdit(edits, id) }
    }
  }
}

/**
 * Keep a local cache of pending/completed edits.
 *
 * To edit a workflow, call `handleWorkflowChanging(id, { public: true })` and
 * then submit the API request. When the request completes, call
 * `handleWorkflowChanged(id)`.
 *
 * This isn't theoretically perfect: if edit B reaches the server before edit A,
 * then we'll locally assume edit A succeeded, even if remotely edit B came
 * last. TODO serialize requests or implement more complex logic.
 */
function useWorkflowEdits (workflows) {
  const [state, dispatch] = React.useReducer(reduceWorkflowEdit, { adds: [], edits: {} })

  const handleWorkflowChanging = React.useCallback((id, changes) => {
    dispatch({ type: 'begin-edit', payload: { id, changes } })
  }, [dispatch])
  const handleWorkflowChanged = React.useCallback(id => {
    dispatch({ type: 'end-edit', payload: { id } })
  }, [dispatch])
  const handleWorkflowDuplicating = React.useCallback(id => {
    dispatch({ type: 'begin-duplicate', payload: { id } })
  }, [dispatch])
  const handleWorkflowDuplicated = React.useCallback((id, workflow) => {
    dispatch({ type: 'end-duplicate', payload: { id, workflow } })
  }, [dispatch])

  const { adds, edits } = state
  const editedWorkflows = adds.concat(workflows).map(workflow => {
    const entry = edits[workflow.id]
    return entry
      ? { ...workflow, ...entry.changes, nPendingChanges: entry.nPendingChanges }
      : workflow
  }).filter(workflow => !workflow.isDeleted || workflow.nPendingChanges > 0)

  return {
    editedWorkflows,
    handleWorkflowChanging,
    handleWorkflowChanged,
    handleWorkflowDuplicating,
    handleWorkflowDuplicated
  }
}

export default function OwnedWorkflowsMain (props) {
  const { api, workflows } = props

  const {
    editedWorkflows,
    handleWorkflowChanging,
    handleWorkflowChanged,
    handleWorkflowDuplicating,
    handleWorkflowDuplicated
  } = useWorkflowEdits(workflows)

  const handleClickDeleteWorkflow = React.useCallback(id => {
    if (!confirm(
      t({ id: 'js.Workflows.delete.permanentyDeleteWarning', message: 'Permanently delete this workflow?' })
    )) {
      return
    }

    handleWorkflowChanging(id, { isDeleted: true })
    api.deleteWorkflow(id).then(() => handleWorkflowChanged(id))
  }, [api, handleWorkflowChanging, handleWorkflowChanged])

  const handleClickDuplicateWorkflow = React.useCallback(id => {
    handleWorkflowDuplicating(id, {})
    api.duplicateWorkflow(id).then(json => {
      handleWorkflowDuplicated(id, json)
    })
  }, [api, handleWorkflowDuplicating, handleWorkflowDuplicated])

  return (
    <main className='workflows'>
      <header>
        <h1><Trans id='js.Workflows.WorkflowLists.nav.myWorkflows'>My workflows</Trans></h1>
      </header>
      {editedWorkflows.length ? (
        <WorkflowList
          className='owned'
          workflows={editedWorkflows}
          onClickDeleteWorkflow={handleClickDeleteWorkflow}
          onClickDuplicateWorkflow={handleClickDuplicateWorkflow}
          apiForShareModal={api}
          onWorkflowChanging={handleWorkflowChanging}
          onWorkflowChanged={handleWorkflowChanged}
        />
      ) : (
        <CreateWorkflowButton>
          <Trans id='js.Workflows.WorkflowLists.createYourFirtsWorkflow.button'>Create your first workflow</Trans>
        </CreateWorkflowButton>
      )}
    </main>
  )
}
OwnedWorkflowsMain.propTypes = {
  workflows: WorkflowListPropType.isRequired,
  api: PropTypes.shape({
    deleteWorkflow: PropTypes.func.isRequired, // func(id) => Promise[null]
    duplicateWorkflow: PropTypes.func.isRequired, // func(id) => Promise[{ id, name }]
    updateAclEntry: PropTypes.func.isRequired, // func(id, email, canEdit) => Promise[null]
    deleteAclEntry: PropTypes.func.isRequired, // func(id, email) => Promise[null]
    setWorkflowPublic: PropTypes.func.isRequired // func(id, isPublic) => Promise[null]
  }).isRequired
}
