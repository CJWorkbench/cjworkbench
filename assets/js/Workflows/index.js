/* globals confirm */
// Elements of /workflows. Navbar plus a list

import React from 'react'
import Navbar from './Navbar'
import PropTypes from 'prop-types'
import ShareModal from '../ShareModal/Modal' // _not_ the Redux-connected component, 'ShareModal'
import { logUserEvent } from '../utils'
import CreateWorkflowButton from './CreateWorkflowButton'
import WorkflowLists from './WorkflowLists'
import { WorkflowListPropType } from './WorkflowList'
import { t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

export class Workflows extends React.Component {
  static propTypes = {
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    }),
    api: PropTypes.shape({
      deleteWorkflow: PropTypes.func.isRequired, // func(id) => Promise[null]
      duplicateWorkflow: PropTypes.func.isRequired // func(id) => Promise[{ id, name }]
    }).isRequired,
    workflows: PropTypes.shape({
      owned: WorkflowListPropType.isRequired,
      shared: WorkflowListPropType.isRequired,
      templates: WorkflowListPropType.isRequired
    }).isRequired,
    user: PropTypes.shape({ id: PropTypes.number.isRequired }) // null/undefined if logged out
  }

  state = {
    workflows: this.props.workflows, // we'll be editing this
    shareModalWorkflowId: null
  }

  get allWorkflows () {
    const { workflows } = this.state
    return [].concat(...Object.keys(workflows).map(k => workflows[k]))
  }

  workflowIdToTabName = (workflowId) => {
    const { workflows } = this.state
    for (const tab in workflows) {
      if (workflows[tab].findIndex(w => w.id === workflowId) !== -1) return tab
    }
    return null
  }

  openShareModal = (workflowId) => {
    this.setState({ shareModalWorkflowId: workflowId })
  }

  handleCloseShareModal = () => {
    this.setState({ shareModalWorkflowId: null })
  }

  logShare = (type) => {
    logUserEvent('Share workflow ' + type)
  }

  renderShareModal = () => {
    const { shareModalWorkflowId } = this.state

    if (shareModalWorkflowId === null) return null

    const workflow = this.allWorkflows.find(w => w.id === shareModalWorkflowId)
    if (!workflow) return null

    const url = `${window.origin}/workflows/${workflow.id}`

    return (
      <ShareModal
        url={url}
        acl={workflow.acl}
        ownerEmail={workflow.owner_email}
        workflowId={workflow.id}
        isReadOnly={!workflow.is_owner}
        isPublic={workflow.public}
        setIsPublic={this.setIsPublic}
        updateAclEntry={this.updateAclEntry}
        deleteAclEntry={this.deleteAclEntry}
        logShare={this.logShare}
        onClickClose={this.handleCloseShareModal}
      />
    )
  }

  // Ask the user if they really wanna do this. If sure, post DELETE to server
  deleteWorkflow = (workflowId) => {
    const tabName = this.workflowIdToTabName(workflowId)
    if (!tabName) return

    if (!confirm(this.props.i18n._(t('js.Workflows.delete.permanentyDeleteWarning')`Permanently delete this workflow?`))) return

    this.props.api.deleteWorkflow(workflowId)
      .then(() => {
        this.setState({
          workflows: {
            ...this.state.workflows,
            [tabName]: this.state.workflows[tabName].filter(w => w.id !== workflowId)
          }
        })
      })
  }

  duplicateWorkflow = (id) => {
    return this.props.api.duplicateWorkflow(id)
      .then(json => {
        // Add to beginning of owned list then set activeTab to owned
        this.setState({
          workflows: {
            ...this.state.workflows,
            owned: [json, ...this.state.workflows.owned]
          }
        })
      })
  }

  /**
   * Change properties of a Workflow in this.state.
   *
   * Calls setState().
   */
  _updateStateWorkflow = (workflowId, assignValues) => {
    const { workflows } = this.state
    // Find workflow category
    const category = Object.keys(workflows)
      .find(k => workflows[k].findIndex(w => w.id === workflowId) !== -1)
    // Copy its workflows, assigning `assignValues` to the one with ID `workflowId`
    const categoryWorkflows = workflows[category]
      .map(w => w.id === workflowId ? { ...w, ...assignValues } : w)

    const newWorkflows = {
      ...workflows,
      [category]: categoryWorkflows
    }
    this.setState({ workflows: newWorkflows })
  }

  setIsPublic = (isPublic) => {
    // Logic copied from ShareModal/actions
    const workflowId = this.state.shareModalWorkflowId
    this._updateStateWorkflow(workflowId, { public: isPublic })
    this.props.api.setWorkflowPublic(workflowId, isPublic) // ignoring whether it works
  }

  updateAclEntry = (email, canEdit) => {
    // Logic copied from ShareModal/actions
    const { workflows } = this.state
    const workflowId = this.state.shareModalWorkflowId
    // Find workflow
    const workflow = Object.keys(workflows)
      .flatMap(k => workflows[k])
      .find(w => w.id === workflowId)
    const acl = workflow.acl.slice() // shallow copy

    let index = acl.findIndex(entry => entry.email === email)
    if (index === -1) index = acl.length

    // overwrite or append the specified ACL entry
    acl[index] = { email, canEdit }

    acl.sort((a, b) => a.email.localeCompare(b.email))

    this._updateStateWorkflow(workflowId, { acl })
    this.props.api.updateAclEntry(workflowId, email, canEdit) // ignoring whether it works
  }

  deleteAclEntry = (email) => {
    // Logic copied from ShareModal/actions
    const { workflows } = this.state
    const workflowId = this.state.shareModalWorkflowId
    // Find workflow
    const workflow = Object.keys(workflows)
      .flatMap(k => workflows[k])
      .find(w => w.id === workflowId)
    const acl = workflow.acl.filter(entry => entry.email !== email)

    this._updateStateWorkflow(workflowId, { acl })
    this.props.api.deleteAclEntry(workflowId, email) // ignoring whether it works
  }

  render () {
    const { workflows } = this.state
    const { user, i18n } = this.props

    return (
      <div className='workflows-page'>
        <Navbar user={user} />
        <a href='/lessons/' className='lesson-banner mx-auto'>
          <div>
            <div className='content-1'>{i18n._(/* i18n: This should be all-caps for styling reasons */t('js.Workflows.new')`NEW`)}</div>
            <div className='d-flex'>
              <span className='icon-star' />
              <div className=' title-1 '>{i18n._(/* i18n: This should be all-caps for styling reasons */t('js.Workflows.training.title')`TRAINING`)}</div>
            </div>
          </div>
          <p>{i18n._(t('js.Workflows.learnHowToWorkWithData')`Learn how to work with data without coding`)}</p>
        </a>
        <CreateWorkflowButton>
          {i18n._(t('js.Workflows.createWorkflowButton')`Create Workflow`)}
        </CreateWorkflowButton>
        <WorkflowLists
          workflows={workflows}
          deleteWorkflow={this.deleteWorkflow}
          duplicateWorkflow={this.duplicateWorkflow}
          openShareModal={this.openShareModal}
        />
        {this.renderShareModal()}
      </div>
    )
  }
}

export default withI18n()(Workflows)
