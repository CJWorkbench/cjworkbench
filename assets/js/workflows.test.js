import React from 'react';
import { mount } from 'enzyme';

import Workflows from './workflows';

it('renders correctly', (done) => {

  var mockWorkflows = [
      {
        "id": 1,
        "name": "Charting"
      },
      {
        "id": 7,
        "name": "Messy data cleanup"
      },
      {
        "id": 8,
        "name": "Document search"
      },
      {
        "id": 9,
        "name": "Visualization"
      },
    ];

  // Initial state, no workflows
  const wrapper = mount( <Workflows /> );
  expect(wrapper).toMatchSnapshot();

  // Now give it some actual workflows to work with
  wrapper.setState( {workflows: mockWorkflows, newWorkflowName: 'foop'} );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('.item-test-class')).toHaveLength(4);

  // Make sure there is a delete button for each workflow
  var buttons = wrapper.find('.button-test-class');
  expect(buttons).toHaveLength(4);

  // Try deleting a workflow
  global.confirm = () => true;                       // pretend the user clicked OK
  global.fetch = () => Promise.resolve({ok: true});  // respond to DELETE with Ok
  buttons.first().simulate('click');

  // We've clicked and now we have to wait for everything to update.
  // We do this with node's setImmediate and Jest's done https://facebook.github.io/jest/docs/asynchronous.html
  setImmediate( () => {
    expect(wrapper.find('.item-test-class')).toHaveLength(3);
    done();
  });

});

