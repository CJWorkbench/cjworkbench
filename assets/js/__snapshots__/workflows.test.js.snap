// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Workflows>
  <div>
    <WorkflowListNavBar>
      <div>
        <nav
          className="navbar"
        >
          <h1
            className="navbar-brand mb-0 mr-auto title-3"
          >
            CJ Workbench
          </h1>
          <WfHamburgerMenu>
            <UncontrolledDropdown>
              <Dropdown
                isOpen={false}
                tag="div"
                toggle={[Function]}
              >
                <div
                  className="dropdown"
                >
                  <DropdownToggle
                    aria-haspopup={true}
                    color="secondary"
                    data-toggle="dropdown"
                  >
                    <Button
                      aria-expanded={false}
                      aria-haspopup="true"
                      className=""
                      color="secondary"
                      data-toggle="dropdown"
                      onClick={[Function]}
                      tag="button"
                    >
                      <button
                        aria-expanded={false}
                        aria-haspopup="true"
                        className="btn btn-secondary"
                        data-toggle="dropdown"
                        onClick={[Function]}
                        type="button"
                      >
                        ☰
                      </button>
                    </Button>
                  </DropdownToggle>
                  <DropdownMenu
                    tag="div"
                  >
                    <div
                      aria-hidden={true}
                      className="dropdown-menu"
                      role="menu"
                      tabIndex="-1"
                    >
                      <DropdownItem
                        className="dropdown-menu-item"
                        href="http://blog.cjworkbench.org"
                        tag="a"
                        toggle={true}
                      >
                        <a
                          className="dropdown-menu-item dropdown-item"
                          href="http://blog.cjworkbench.org"
                          onClick={[Function]}
                          tabIndex="0"
                        >
                          <span
                            className="t-d-gray content-3 ml-3"
                          >
                            Help
                          </span>
                        </a>
                      </DropdownItem>
                      <DropdownItem
                        className="dropdown-menu-item"
                        href="/account/logout"
                        tag="a"
                        toggle={true}
                      >
                        <a
                          className="dropdown-menu-item dropdown-item"
                          href="/account/logout"
                          onClick={[Function]}
                          tabIndex="0"
                        >
                          <span
                            className="t-d-gray content-3 ml-3"
                          >
                            Logout
                          </span>
                        </a>
                      </DropdownItem>
                    </div>
                  </DropdownMenu>
                </div>
              </Dropdown>
            </UncontrolledDropdown>
          </WfHamburgerMenu>
        </nav>
      </div>
    </WorkflowListNavBar>
    <div
      className="container workflows-container"
    >
      <div
        className="row justify-content-md-center"
      >
        <div
          className="col col-lg-2"
        />
        <div
          className="col-12 col-md-auto"
        >
          <div
            className="input-group"
          >
            <span
              className="input-group-btn"
            >
              <div
                className="button-blue action-button new-workflow-button"
                onClick={[Function]}
              >
                New
              </div>
            </span>
          </div>
        </div>
        <div
          className="col col-lg-2"
        />
      </div>
      <div
        className="card w-75 mx-auto workflows-list"
      >
        <div
          className="card-block"
        >
          <h3
            className="card-title title-2 t-d-gray ml-4 mb-3"
          >
            Workflows
          </h3>
          <div
            className=""
          />
        </div>
      </div>
    </div>
  </div>
</Workflows>
`;

exports[`renders correctly 2`] = `
<Workflows>
  <div>
    <WorkflowListNavBar>
      <div>
        <nav
          className="navbar"
        >
          <h1
            className="navbar-brand mb-0 mr-auto title-3"
          >
            CJ Workbench
          </h1>
          <WfHamburgerMenu>
            <UncontrolledDropdown>
              <Dropdown
                isOpen={false}
                tag="div"
                toggle={[Function]}
              >
                <div
                  className="dropdown"
                >
                  <DropdownToggle
                    aria-haspopup={true}
                    color="secondary"
                    data-toggle="dropdown"
                  >
                    <Button
                      aria-expanded={false}
                      aria-haspopup="true"
                      className=""
                      color="secondary"
                      data-toggle="dropdown"
                      onClick={[Function]}
                      tag="button"
                    >
                      <button
                        aria-expanded={false}
                        aria-haspopup="true"
                        className="btn btn-secondary"
                        data-toggle="dropdown"
                        onClick={[Function]}
                        type="button"
                      >
                        ☰
                      </button>
                    </Button>
                  </DropdownToggle>
                  <DropdownMenu
                    tag="div"
                  >
                    <div
                      aria-hidden={true}
                      className="dropdown-menu"
                      role="menu"
                      tabIndex="-1"
                    >
                      <DropdownItem
                        className="dropdown-menu-item"
                        href="http://blog.cjworkbench.org"
                        tag="a"
                        toggle={true}
                      >
                        <a
                          className="dropdown-menu-item dropdown-item"
                          href="http://blog.cjworkbench.org"
                          onClick={[Function]}
                          tabIndex="0"
                        >
                          <span
                            className="t-d-gray content-3 ml-3"
                          >
                            Help
                          </span>
                        </a>
                      </DropdownItem>
                      <DropdownItem
                        className="dropdown-menu-item"
                        href="/account/logout"
                        tag="a"
                        toggle={true}
                      >
                        <a
                          className="dropdown-menu-item dropdown-item"
                          href="/account/logout"
                          onClick={[Function]}
                          tabIndex="0"
                        >
                          <span
                            className="t-d-gray content-3 ml-3"
                          >
                            Logout
                          </span>
                        </a>
                      </DropdownItem>
                    </div>
                  </DropdownMenu>
                </div>
              </Dropdown>
            </UncontrolledDropdown>
          </WfHamburgerMenu>
        </nav>
      </div>
    </WorkflowListNavBar>
    <div
      className="container workflows-container"
    >
      <div
        className="row justify-content-md-center"
      >
        <div
          className="col col-lg-2"
        />
        <div
          className="col-12 col-md-auto"
        >
          <div
            className="input-group"
          >
            <span
              className="input-group-btn"
            >
              <div
                className="button-blue action-button new-workflow-button"
                onClick={[Function]}
              >
                New
              </div>
            </span>
          </div>
        </div>
        <div
          className="col col-lg-2"
        />
      </div>
      <div
        className="card w-75 mx-auto workflows-list"
      >
        <div
          className="card-block"
        >
          <h3
            className="card-title title-2 t-d-gray ml-4 mb-3"
          >
            Workflows
          </h3>
          <div
            className=""
          >
            <div
              className="card card-block item-test-class workflow-in-list"
            >
              <a
                href="/workflows/1"
              >
                <div
                  className="d-flex justify-content-between"
                >
                  <span
                    className="t-d-gray title-4"
                  >
                    Charting
                  </span>
                  <div
                    className="menu-test-class"
                    onClick={[Function]}
                  >
                    <WfContextMenu
                      deleteWorkflow={[Function]}
                    >
                      <UncontrolledDropdown>
                        <Dropdown
                          isOpen={false}
                          tag="div"
                          toggle={[Function]}
                        >
                          <div
                            className="dropdown"
                          >
                            <DropdownToggle
                              aria-haspopup={true}
                              className="context-button"
                              color="secondary"
                              data-toggle="dropdown"
                            >
                              <Button
                                aria-expanded={false}
                                aria-haspopup="true"
                                className="context-button"
                                color="secondary"
                                data-toggle="dropdown"
                                onClick={[Function]}
                                tag="button"
                              >
                                <button
                                  aria-expanded={false}
                                  aria-haspopup="true"
                                  className="context-button btn btn-secondary"
                                  data-toggle="dropdown"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <div
                                    className="button-icon icon-more"
                                  />
                                </button>
                              </Button>
                            </DropdownToggle>
                            <DropdownMenu
                              tag="div"
                            >
                              <div
                                aria-hidden={true}
                                className="dropdown-menu"
                                role="menu"
                                tabIndex="-1"
                              >
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-share "
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Share
                                    </span>
                                  </button>
                                </DropdownItem>
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-bin"
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Delete
                                    </span>
                                  </button>
                                </DropdownItem>
                              </div>
                            </DropdownMenu>
                          </div>
                        </Dropdown>
                      </UncontrolledDropdown>
                    </WfContextMenu>
                  </div>
                </div>
              </a>
            </div>
            <div
              className="card card-block item-test-class workflow-in-list"
            >
              <a
                href="/workflows/7"
              >
                <div
                  className="d-flex justify-content-between"
                >
                  <span
                    className="t-d-gray title-4"
                  >
                    Messy data cleanup
                  </span>
                  <div
                    className="menu-test-class"
                    onClick={[Function]}
                  >
                    <WfContextMenu
                      deleteWorkflow={[Function]}
                    >
                      <UncontrolledDropdown>
                        <Dropdown
                          isOpen={false}
                          tag="div"
                          toggle={[Function]}
                        >
                          <div
                            className="dropdown"
                          >
                            <DropdownToggle
                              aria-haspopup={true}
                              className="context-button"
                              color="secondary"
                              data-toggle="dropdown"
                            >
                              <Button
                                aria-expanded={false}
                                aria-haspopup="true"
                                className="context-button"
                                color="secondary"
                                data-toggle="dropdown"
                                onClick={[Function]}
                                tag="button"
                              >
                                <button
                                  aria-expanded={false}
                                  aria-haspopup="true"
                                  className="context-button btn btn-secondary"
                                  data-toggle="dropdown"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <div
                                    className="button-icon icon-more"
                                  />
                                </button>
                              </Button>
                            </DropdownToggle>
                            <DropdownMenu
                              tag="div"
                            >
                              <div
                                aria-hidden={true}
                                className="dropdown-menu"
                                role="menu"
                                tabIndex="-1"
                              >
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-share "
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Share
                                    </span>
                                  </button>
                                </DropdownItem>
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-bin"
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Delete
                                    </span>
                                  </button>
                                </DropdownItem>
                              </div>
                            </DropdownMenu>
                          </div>
                        </Dropdown>
                      </UncontrolledDropdown>
                    </WfContextMenu>
                  </div>
                </div>
              </a>
            </div>
            <div
              className="card card-block item-test-class workflow-in-list"
            >
              <a
                href="/workflows/8"
              >
                <div
                  className="d-flex justify-content-between"
                >
                  <span
                    className="t-d-gray title-4"
                  >
                    Document search
                  </span>
                  <div
                    className="menu-test-class"
                    onClick={[Function]}
                  >
                    <WfContextMenu
                      deleteWorkflow={[Function]}
                    >
                      <UncontrolledDropdown>
                        <Dropdown
                          isOpen={false}
                          tag="div"
                          toggle={[Function]}
                        >
                          <div
                            className="dropdown"
                          >
                            <DropdownToggle
                              aria-haspopup={true}
                              className="context-button"
                              color="secondary"
                              data-toggle="dropdown"
                            >
                              <Button
                                aria-expanded={false}
                                aria-haspopup="true"
                                className="context-button"
                                color="secondary"
                                data-toggle="dropdown"
                                onClick={[Function]}
                                tag="button"
                              >
                                <button
                                  aria-expanded={false}
                                  aria-haspopup="true"
                                  className="context-button btn btn-secondary"
                                  data-toggle="dropdown"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <div
                                    className="button-icon icon-more"
                                  />
                                </button>
                              </Button>
                            </DropdownToggle>
                            <DropdownMenu
                              tag="div"
                            >
                              <div
                                aria-hidden={true}
                                className="dropdown-menu"
                                role="menu"
                                tabIndex="-1"
                              >
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-share "
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Share
                                    </span>
                                  </button>
                                </DropdownItem>
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-bin"
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Delete
                                    </span>
                                  </button>
                                </DropdownItem>
                              </div>
                            </DropdownMenu>
                          </div>
                        </Dropdown>
                      </UncontrolledDropdown>
                    </WfContextMenu>
                  </div>
                </div>
              </a>
            </div>
            <div
              className="card card-block item-test-class workflow-in-list"
            >
              <a
                href="/workflows/9"
              >
                <div
                  className="d-flex justify-content-between"
                >
                  <span
                    className="t-d-gray title-4"
                  >
                    Visualization
                  </span>
                  <div
                    className="menu-test-class"
                    onClick={[Function]}
                  >
                    <WfContextMenu
                      deleteWorkflow={[Function]}
                    >
                      <UncontrolledDropdown>
                        <Dropdown
                          isOpen={false}
                          tag="div"
                          toggle={[Function]}
                        >
                          <div
                            className="dropdown"
                          >
                            <DropdownToggle
                              aria-haspopup={true}
                              className="context-button"
                              color="secondary"
                              data-toggle="dropdown"
                            >
                              <Button
                                aria-expanded={false}
                                aria-haspopup="true"
                                className="context-button"
                                color="secondary"
                                data-toggle="dropdown"
                                onClick={[Function]}
                                tag="button"
                              >
                                <button
                                  aria-expanded={false}
                                  aria-haspopup="true"
                                  className="context-button btn btn-secondary"
                                  data-toggle="dropdown"
                                  onClick={[Function]}
                                  type="button"
                                >
                                  <div
                                    className="button-icon icon-more"
                                  />
                                </button>
                              </Button>
                            </DropdownToggle>
                            <DropdownMenu
                              tag="div"
                            >
                              <div
                                aria-hidden={true}
                                className="dropdown-menu"
                                role="menu"
                                tabIndex="-1"
                              >
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-share "
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Share
                                    </span>
                                  </button>
                                </DropdownItem>
                                <DropdownItem
                                  className="dropdown-menu-item"
                                  onClick={[Function]}
                                  tag="button"
                                  toggle={true}
                                >
                                  <button
                                    className="dropdown-menu-item dropdown-item"
                                    onClick={[Function]}
                                    tabIndex="0"
                                    type="button"
                                  >
                                    <i
                                      className="icon-bin"
                                    />
                                    <span
                                      className="t-d-gray content-3 ml-3"
                                    >
                                      Delete
                                    </span>
                                  </button>
                                </DropdownItem>
                              </div>
                            </DropdownMenu>
                          </div>
                        </Dropdown>
                      </UncontrolledDropdown>
                    </WfContextMenu>
                  </div>
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</Workflows>
`;
